rust/foxglove/src/stream.rs

```diff
@@ -0,0 +1,266 @@
+        }
+
+        Ok(())
+    }
+
+    /// Get the current size of the buffer.
+    ///
+    /// This can be used in conjunction with [`McapStreamHandle::flush`] to ensure the buffer does
+    /// not grow unbounded.
+    pub fn buffer_size(&mut self) -> usize {
+        self.buffer.0.lock().buffer.len()
+    }
+
+    /// Flush the buffer from the MCAP writer to the [`McapStream`].
+    ///
+    /// This method returns a future that will wait until the [`McapStream`] has capacity for the
+    /// flushed buffer.
+    pub async fn flush(&mut self) -> Result<(), FoxgloveError> {
+        Self::flush_shared_buffer(&mut self.sender, &mut self.buffer).await?;
+        Ok(())
+    }
+}
+
+impl Drop for McapStreamHandle {
+    fn drop(&mut self) {
+        if let Some(writer) = self.writer.take() {
+            if let Err(e) = writer.close() {
+                tracing::warn!("{e}");
+            }
+        }
+
+        let mut inner = self.buffer.0.lock();
+        let buffer = inner.buffer.split();
+
+        if !buffer.is_empty() {
+            // When the handle is dropped try and send the final buffer. If the channel is full or
+            // closed log as a warning.
+            if let Err(e) = self.sender.try_send(buffer) {
+                tracing::warn!("{e}");
+            }
+        }
+    }
+}
+
+/// A stream of MCAP bytes from a writer.
+pub struct McapStream {
+    receiver: TokioReceiver<BytesMut>,
+}
+
+impl Stream for McapStream {
+    type Item = Bytes;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let Some(bytes) = ready!(self.receiver.poll_recv(cx)) else {
+            return Poll::Ready(None);
+        };
+
+        Poll::Ready(Some(bytes.freeze()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use futures::StreamExt;
+    use std::convert::Infallible;
+
+    use crate::Encode;
+
+    use super::*;
+
+    struct Message {
+        data: f64,
+    }
+
+    impl Encode for Message {
+        type Error = Infallible;
+
+        fn get_schema() -> Option<crate::Schema> {
+            None
+        }
+
+        fn get_message_encoding() -> String {
+            "foo".to_string()
+        }
+
+        fn encode(&self, buf: &mut impl bytes::BufMut) -> Result<(), Self::Error> {
+            buf.put_f64(self.data);
+            Ok(())
+        }
+    }
+
+    #[tokio::test]
+    async fn test_write_to_stream() {
+        let (mut handle, mut stream) = create_mcap_stream();
+
+        let channel = handle.channel_builder("/topic").build::<Message>();
+
+        tokio::spawn(async move {
+            for i in 0..100 {
+                channel.log(&Message { data: i as f64 });
+                handle.flush().await.unwrap();
+            }
+
+            handle.close().await.unwrap();
+        });
+
+        let mut out = vec![];
+
+        while let Some(bytes) = stream.next().await {
```

Comment:
I think this test serves as useful example code and needs comments to explain to readers exactly what's happening and where you'd expect to use this stream. Maybe explain that the stream can be used as an axum response body.